Sources/AsyncHTTPClient/ConnectionPool/HTTP2/HTTP2ClientRequestHandler.swift

```diff
@@ -0,0 +1,330 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Logging
+import NIO
+import NIOHTTP1
+import NIOHTTP2
+
+final class HTTP2ClientRequestHandler: ChannelDuplexHandler {
+    typealias OutboundIn = HTTPExecutableRequest
+    typealias OutboundOut = HTTPClientRequestPart
+    typealias InboundIn = HTTPClientResponsePart
+
+    private let eventLoop: EventLoop
+
+    private var state: HTTPRequestStateMachine = .init(isChannelWritable: false) {
+        didSet {
+            self.eventLoop.assertInEventLoop()
+        }
+    }
+
+    /// while we are in a channel pipeline, this context can be used.
+    private var channelContext: ChannelHandlerContext?
+
+    private var request: HTTPExecutableRequest? {
+        didSet {
+            if let newRequest = self.request, let idleReadTimeout = newRequest.idleReadTimeout {
+                self.idleReadTimeoutStateMachine = .init(timeAmount: idleReadTimeout)
+            } else {
+                self.idleReadTimeoutStateMachine = nil
+            }
+        }
+    }
+
+    private var idleReadTimeoutStateMachine: IdleReadStateMachine?
+    private var idleReadTimeoutTimer: Scheduled<Void>?
+
+    init(eventLoop: EventLoop) {
+        self.eventLoop = eventLoop
+    }
+
+    func handlerAdded(context: ChannelHandlerContext) {
```

Comment:
Can I propose validating that the event loop associated with our channel and the loop passed to us in `init` are the same here?